### PR TITLE
PLT-312: Unit test for VSphere cluster

### PR DESCRIPTION
**Pull request by SivaanandM**

### CODE REVIEW

File: spectrocloud/resource_cluster_vsphere_test.go [hunk 1]
```diff
@@ -0,0 +1,741 @@
...
+	assert.Equal("spectrocloud.dev", vSphereSchema.Spec.CloudConfig.ControlPlaneEndpoint.DdnsSearchDomain)
+	assert.Equal("DDNS", vSphereSchema.Spec.CloudConfig.ControlPlaneEndpoint.Type)
+	assert.Equal("Datacenter", vSphereSchema.Spec.CloudConfig.Placement.Datacenter)
+	assert.Equal("sc_test/terraform", vSphereSchema.Spec.CloudConfig.Placement.Folder)
+	assert.Equal("ssh-rsa AAAAB3NzaC1y", vSphereSchema.Spec.CloudConfig.SSHKeys[0])
+	assert.Equal(false, vSphereSchema.Spec.CloudConfig.StaticIP)
+
+	// Verifying Master pool attributes
+	assert.Equal("master-pool", *vSphereSchema.Spec.Machinepoolconfig[0].PoolConfig.Name)
+	assert.Equal(true, vSphereSchema.Spec.Machinepoolconfig[0].PoolConfig.IsControlPlane)
```
Comment (nikchern) [inline spectrocloud/resource_cluster_vsphere_test.go:155]: Is it possible that pools will change order? May be better to rewrite assert no matter on what is the order.

File: spectrocloud/resource_cluster_vsphere_test.go [hunk 2]
```diff
@@ -0,0 +1,741 @@
...
+			*numCpu = 2
+			network := new(string)
+			*network = "test-net"
+			mTaint = append(mTaint, &models.V1Taint{
+				Effect:    "start",
+				Key:       "owner",
+				TimeAdded: models.V1Time{},
+				Value:     "siva",
+			})
+			var place []*models.V1VspherePlacementConfig
```
Comment (nikchern) [inline spectrocloud/resource_cluster_vsphere_test.go:568]: placement is a better variable name.

File: spectrocloud/resource_cluster_vsphere_test.go [hunk 3]
```diff
@@ -0,0 +1,741 @@
...
+				Name:                 "",
+				Placements:           nil,
+				Size:                 0,
+				Taints:               mTaint,
+				UpdateStrategy: &models.V1UpdateStrategy{
+					Type: "",
+				},
+				UseControlPlaneAsWorker: false,
+			})
+			cloudConfig := &models.V1VsphereCloudConfig{
```
Comment (nikchern) [inline spectrocloud/resource_cluster_vsphere_test.go:272]: let's reuse it across tests.

File: spectrocloud/resource_cluster_vsphere_test.go [hunk 4]
```diff
@@ -0,0 +1,741 @@
...
+			network := new(string)
+			*network = "test-net"
+			mTaint = append(mTaint, &models.V1Taint{
+				Effect:    "start",
+				Key:       "owner",
+				TimeAdded: models.V1Time{},
+				Value:     "siva",
+			})
+			var place []*models.V1VspherePlacementConfig
+			place = append(place, &models.V1VspherePlacementConfig{
```
Comment (nikchern) [inline spectrocloud/resource_cluster_vsphere_test.go:569]: can this be reused across tests?

File: spectrocloud/resource_cluster_vsphere_test.go [hunk 5]
```diff
@@ -0,0 +1,741 @@
...
+		GetCloudConfigVsphereFn: func(uid string) (*models.V1VsphereCloudConfig, error) {
+			var mTaint []*models.V1Taint
+			mTaint = append(mTaint, &models.V1Taint{
+				Effect:    "start",
+				Key:       "owner",
+				TimeAdded: models.V1Time{},
+				Value:     "siva",
+			})
+			var mPool []*models.V1VsphereMachinePoolConfig
+			mPool = append(mPool, &models.V1VsphereMachinePoolConfig{
```
Comment (nikchern) [inline spectrocloud/resource_cluster_vsphere_test.go:250]: may be also can be reused

**Review by nikchern — CHANGES_REQUESTED** [review]

File: spectrocloud/resource_cluster_vsphere_test.go (hunk 1 above)
Comment (SivaanandM) [reply spectrocloud/resource_cluster_vsphere_test.go]: Handled change order cases for master and worker pool

File: spectrocloud/resource_cluster_vsphere_test.go (hunk 3 above)
Comment (SivaanandM) [reply spectrocloud/resource_cluster_vsphere_test.go]: return reuse method

File: spectrocloud/resource_cluster_vsphere_test.go (hunk 2 above)
Comment (SivaanandM) [reply spectrocloud/resource_cluster_vsphere_test.go]: changed

File: spectrocloud/resource_cluster_vsphere_test.go (hunk 4 above)
Comment (SivaanandM) [reply spectrocloud/resource_cluster_vsphere_test.go]: added reuse method

File: spectrocloud/resource_cluster_vsphere_test.go (hunk 5 above)
Comment (SivaanandM) [reply spectrocloud/resource_cluster_vsphere_test.go]: added reuse method

**Review by nikchern — APPROVED** [review]

**Pull request merged**